File: src/morphic/inspector/scanner.py
"""
Background scan job management for the inspector module.

Handles EXIF scanning and integrity checking in background threads.
"""

from __future__ import annotations

import logging
import os
import threading
import time
import uuid
from dataclasses import dataclass, field

from morphic.shared.constants import ALL_EXTENSIONS, IMAGE_EXTENSIONS
from morphic.shared.utils import (
    find_files_by_extension,
    format_duration,
    is_image,
    is_video,
)

logger = logging.getLogger(__name__)


@dataclass
class ScanJob:
    """Represents a running or completed inspector job."""

    id: str
    folder: str
    mode: str  # "exif" or "integrity"
    status: str = "pending"
    progress: float = 0.0
    message: str = ""
    error: str | None = None
    results: list[dict] = field(default_factory=list)
    total_files: int = 0
    processed_files: int = 0
    started_at: float = 0.0
    finished_at: float = 0.0


# ── Job Registry ───────────────────────────────────────────────────────────

_jobs: dict[str, ScanJob] = {}
_lock = threading.Lock()


def get_job(job_id: str) -> ScanJob | None:
    """Retrieve a scan job by ID."""
    with _lock:
        return _jobs.get(job_id)


def start_job(folder: str, mode: str) -> str:
    """Create and launch a new inspector job. Returns the job ID."""
    job_id = str(uuid.uuid4())[:8]
    job = ScanJob(id=job_id, folder=folder, mode=mode)
    with _lock:
        _jobs[job_id] = job

    thread = threading.Thread(target=_run_scan, args=(job,), daemon=True)
    thread.start()
    return job_id


def _run_scan(job: ScanJob) -> None:
    """Execute the inspector scan in a background thread."""
    try:
        job.status = "scanning"
        job.started_at = time.time()

        # Determine extensions to look for
        extensions = IMAGE_EXTENSIONS if job.mode == "exif" else ALL_EXTENSIONS
        job.message = f"Scanning folder: {job.folder}"
        paths = find_files_by_extension(job.folder, extensions)
        job.total_files = len(paths)

        if not paths:
            job.status = "done"
            job.progress = 1.0
            job.finished_at = time.time()
            job.message = "No files found."
            return

        job.status = "processing"

        if job.mode == "exif":
            _scan_exif(job, paths)
        else:
            _scan_integrity(job, paths)

        job.status = "done"
        job.progress = 1.0
        job.finished_at = time.time()
        elapsed = job.finished_at - job.started_at
        job.message = (
            f"Done! Processed {job.processed_files} files "
            f"in {format_duration(elapsed)}."
        )

    except Exception as e:
        logger.exception("Inspector scan failed")
        job.status = "error"
        job.error = str(e)
        job.message = f"Error: {e}"
        job.finished_at = time.time()


def _scan_exif(job: ScanJob, paths: list[str]) -> None:
    """Read EXIF from all image files."""
    from morphic.inspector.exif import read_exif

    for i, path in enumerate(paths):
        if not is_image(path):
            continue
        try:
            exif = read_exif(path)
            job.results.append(
                {
                    "path": path,
                    "filename": os.path.basename(path),
                    "directory": os.path.dirname(path),
                    "exif": exif,
                    "has_exif": bool(exif),
                    "has_gps": "_gps_lat" in exif,
                }
            )
        except Exception as e:
            job.results.append(
                {
                    "path": path,
                    "filename": os.path.basename(path),
                    "directory": os.path.dirname(path),
                    "exif": {},
                    "has_exif": False,
                    "has_gps": False,
                    "error": str(e),
                }
            )
        job.processed_files = i + 1
        job.progress = (i + 1) / job.total_files
        job.message = f"Reading EXIF: {i + 1}/{job.total_files}"


def _scan_integrity(job: ScanJob, paths: list[str]) -> None:
    """Check integrity of all media files."""
    from morphic.inspector.integrity import check_image, check_video

    for i, path in enumerate(paths):
        if is_image(path):
            result = check_image(path)
        elif is_video(path):
            result = check_video(path)
        else:
            result = {
                "path": path,
                "valid": False,
                "error": "Unknown file type",
                "type": "unknown",
            }
        result["filename"] = os.path.basename(path)
        result["directory"] = os.path.dirname(path)
        job.results.append(result)
        job.processed_files = i + 1
        job.progress = (i + 1) / job.total_files
        job.message = f"Checking: {i + 1}/{job.total_files}"

File: go/internal/inspector/scanner.go
package inspector

import (
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/exterex/morphic/internal/shared"
)

// ScanJob represents a running or completed inspector job.
type ScanJob struct {
	shared.Job
	mu sync.Mutex

	Folder    string                   `json:"folder"`
	Mode      string                   `json:"mode"` // "exif" or "integrity"
	Results   []map[string]interface{} `json:"results,omitempty"`
	Total     int                      `json:"total"`
	Processed int                      `json:"processed"`
}

var store = shared.NewJobStore[ScanJob]()

func init() {
	store.StartCleanup(30*time.Minute, func(j *ScanJob) time.Time {
		return j.DoneAt
	})
}

// StartJob creates and launches a new inspector job.
func StartJob(folder, mode string) string {
	job := &ScanJob{
		Job:    shared.NewJob(),
		Folder: folder,
		Mode:   mode,
	}
	job.Status = shared.JobStatusRunning
	store.Set(job.ID, job)
	go runScan(job)
	return job.ID
}

// GetJob retrieves a job by ID.
func GetJob(id string) (*ScanJob, bool) {
	return store.Get(id)
}

func runScan(job *ScanJob) {
	defer func() {
		if r := recover(); r != nil {
			job.mu.Lock()
			job.Status = shared.JobStatusFailed
			job.Error = fmt.Sprintf("%v", r)
			job.DoneAt = time.Now()
			job.mu.Unlock()
		}
	}()

	job.mu.Lock()
	job.Message = fmt.Sprintf("Scanning folder: %s", job.Folder)
	job.mu.Unlock()

	// Determine extensions to look for
	var exts map[string]struct{}
	if job.Mode == "exif" {
		exts = shared.ImageExtensions
	} else {
		exts = make(map[string]struct{})
		for k, v := range shared.ImageExtensions {
			exts[k] = v
		}
		for k, v := range shared.VideoExtensions {
			exts[k] = v
		}
	}

	files, err := shared.FindFilesByExtension(job.Folder, exts, shared.ExcludedFolders)
	if err != nil {
		job.mu.Lock()
		job.Status = shared.JobStatusFailed
		job.Error = err.Error()
		job.DoneAt = time.Now()
		job.mu.Unlock()
		return
	}

	job.mu.Lock()
	job.Total = len(files)
	job.mu.Unlock()

	if len(files) == 0 {
		job.mu.Lock()
		job.Status = shared.JobStatusDone
		job.Progress = 1.0
		job.DoneAt = time.Now()
		job.Message = "No files found."
		job.mu.Unlock()
		return
	}

	if job.Mode == "exif" {
		scanExif(job, files)
	} else {
		scanIntegrity(job, files)
	}

	job.mu.Lock()
	job.Status = shared.JobStatusDone
	job.Progress = 1.0
	job.DoneAt = time.Now()
	elapsed := job.DoneAt.Sub(job.StartedAt).Seconds()
	job.Message = fmt.Sprintf("Done! Processed %d files in %s.",
		job.Processed, shared.FormatDuration(elapsed))
	job.mu.Unlock()
}

func scanExif(job *ScanJob, files []shared.FileInfo) {
	for i, f := range files {
		if !shared.IsImage(f.Path) {
			continue
		}

		item := map[string]interface{}{
			"path":      f.Path,
			"filename":  filepath.Base(f.Path),
			"directory": filepath.Dir(f.Path),
			"has_exif":  false,
			"has_gps":   false,
		}

		exifData, err := ReadExif(f.Path)
		if err != nil {
			item["exif"] = ExifData{}
			item["error"] = err.Error()
		} else {
			item["exif"] = exifData
			item["has_exif"] = len(exifData) > 0
			_, hasGPS := exifData["_gps_lat"]
			item["has_gps"] = hasGPS
		}

		job.mu.Lock()
		job.Results = append(job.Results, item)
		job.Processed = i + 1
		job.Progress = float64(i+1) / float64(job.Total)
		job.Message = fmt.Sprintf("Reading EXIF: %d/%d", i+1, job.Total)
		job.mu.Unlock()
	}
}

func scanIntegrity(job *ScanJob, files []shared.FileInfo) {
	for i, f := range files {
		var result IntegrityResult
		if shared.IsImage(f.Path) {
			result = CheckImage(f.Path)
		} else if shared.IsVideo(f.Path) {
			result = CheckVideo(f.Path)
		} else {
			result = IntegrityResult{
				Path:  f.Path,
				Type:  "unknown",
				Error: "Unknown file type",
			}
		}
		result.Filename = filepath.Base(f.Path)
		result.Directory = filepath.Dir(f.Path)

		item := map[string]interface{}{
			"path":           result.Path,
			"valid":          result.Valid,
			"error":          nilIfEmpty(result.Error),
			"size":           result.Size,
			"size_formatted": result.SizeFormatted,
			"width":          result.Width,
			"height":         result.Height,
			"type":           result.Type,
			"filename":       result.Filename,
			"directory":      result.Directory,
		}
		if result.Format != "" {
			item["format"] = result.Format
		}
		if result.Codec != "" {
			item["codec"] = result.Codec
		}
		if result.Duration > 0 {
			item["duration"] = result.Duration
		}

		job.mu.Lock()
		job.Results = append(job.Results, item)
		job.Processed = i + 1
		job.Progress = float64(i+1) / float64(job.Total)
		job.Message = fmt.Sprintf("Checking: %d/%d", i+1, job.Total)
		job.mu.Unlock()
	}
}

func nilIfEmpty(s string) interface{} {
	if s == "" {
		return nil
	}
	return s
}

// Stat checks if a path exists.
func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
